lib/llvmopencl/linker.cpp

```diff
@@ -422,70 +469,56 @@ int link(llvm::Module *Program, const llvm::Module *Lib, std::string &Log,
-  llvm::StringSet<>::iterator di,de;
-  for (di = DeclaredFunctions.begin(), de = DeclaredFunctions.end();
-       di != de; di++) {
-      llvm::StringRef r = di->getKey();
-      if (copy_func_callgraph(r, Lib, Program, vvm)) {
-        Function *f = Program->getFunction(r);
-
-        if (f->getName().equals("__to_local") ||
-            f->getName().equals("__to_global") ||
-            f->getName().equals("__to_private")) {
-
-          // Special handling for the AS cast built-ins: They do not use
-          // type mangling, which complicates the CPU implementation which
-          // sometimes gets all-0 AS input and sometimes not (SPIR-V).
-          // Here, in case the target doesn't define these built-ins in the
-          // bitcode library, we replace the calls directly with address
-          // space casts. This works for uniform address space targets (CPUs),
-          // while the disjoint AS targets can override the implementation
-          // as needed.
-          for (auto *U : f->users()) {
-            if (llvm::CallInst *Call = dyn_cast<llvm::CallInst>(U)) {
-            PointerType *ArgPT =
-                dyn_cast<PointerType>(Call->getArgOperand(0)->getType());
-            PointerType *RetPT =
-                dyn_cast<PointerType>(Call->getFunctionType()->getReturnType());
-            if (ArgPT == nullptr || RetPT == nullptr) {
-              Log.append("Invalid use of operator __to_{local,global,private}");
-              found_all_undefined = false;
-              break;
-            }
-            if (ArgPT->getAddressSpace() == RetPT->getAddressSpace()) {
-              Value *V = Call->getArgOperand(0);
-              Call->replaceAllUsesWith(V);
-              Call->eraseFromParent();
-            } else {
-              llvm::AddrSpaceCastInst *AsCast = new llvm::AddrSpaceCastInst(
-                  Call->getArgOperand(0),
-                  Call->getFunctionType()->getReturnType(),
-                  f->getName() + ".as_cast", Call);
-              Call->replaceAllUsesWith(AsCast);
-              Call->eraseFromParent();
-            }
-            }
-          }
-          continue;
-        }
 
-        if ((f == NULL) || (f->isDeclaration() &&
+  for (DI = DeclaredFunctions.begin(), DE = DeclaredFunctions.end();
+       DI != DE; DI++) {
+    llvm::StringRef FName = DI->getKey();
+    Function *F = Program->getFunction(FName);
+
+        if ((F == NULL) || (F->isDeclaration() &&
                             // A target might want to expose the C99 printf in
                             // case not supporting the OpenCL 1.2 printf.
-                            !f->getName().equals("printf") &&
-                            !f->getName().equals(pocl_sampler_handler) &&
-                            !f->getName().starts_with(llvm_intrins))) {
+                            !F->getName().equals("printf") &&
+                            !F->getName().equals(pocl_sampler_handler) &&
```

Comment:
Use == instead of equals() for string comparisons. equals() is gone in LLVM 19.

Reply:
(this will be fixed in my LLVM 19 PR so there might be a conflict)

Reply:
fixed